### ISSUE TITLE
Discussions about how a third party could provide an alternative Dart formatter for VS Code

### ISSUE THREAD

**Issue by tekert** [Dart-Code/Dart-Code#4130]
**Is your feature request related to a problem? Please describe.**
I think everybody knows by now.

**Describe the solution you'd like**
A setting to enter a path to a custom formatter made in dart (its based on the current one so args should work the same) 

**Describe alternatives you've considered**
Closing my eyes. dindt work.

**Additional context**
A way to specify from a list formatters installed from dart pub global activate would also be great.
Basically a way to swap formatters.
(I already made one formatter compliant with the oficial AST tree from the analizer. Has no settings yet so im using dart_style as a extension, so the arguments are the same)

**Comment by DanTup** [comment]
Formatting isn't done by running a separate process, that would be quite slow. Instead it's handled by the language server.

To build a custom formatter, it would be best to wrap it up as its own VS Code extension and allow users to install and select it in VS Code (VS Code already has a UI for selecting a formatter if there are multiple installed, and this extension has a setting that lets you unregister its formatter).

I added some notes about how I would do this here:

https://github.com/Dart-Code/Dart-Code/issues/914#issuecomment-739474329

If someone gets as far as having a working formatter, I'm happy to provide pointers for how to wrap it up in a VS Code extension.

I'll close this because I don't think adding a setting here is the right way to go about it (so there's no work to do), but feel free to post comments if you build a formatter and need any pointers from me. Be warned that if the issue turned into a thread of complaints about the existing formatter (or decisions made by the Dart team) it will be locked. Thanks!

**Comment by tekert** [comment]
Hi Dan, the formatter is already built, and compiled natively, never made a vscode extension im actually a c++ programer that wanted to test dart a bit and got a custom formatter while i was practicing dart.
So, the customizing the dart LSP is a no go?
And how i would go about wraping it as an vscode extension? Thanks.

**Comment by DanTup** [comment]
@tekert I got a notification for your reply, but looks like you deleted it. Did you figure out what you needed, or still need help?

In the first instance, I would make the formatter able to run as a long-lived process with some JSON API over stdin/stdout, so that the VS Code extension would be able to run it in the background and send formatting requests without needing to spawn a new process on each request.

**Comment by tekert** [comment]
Hi dan, yep, im doing that now, maybe this week ill have something, i was looking at how the Language server of dart does it so as not to reinvent the wheel so to speak. And then figure out how to plug dart code on a vscode ext.

**Comment by DanTup** [comment]
To answer your previous question about LSP, it's not possible to have the LSP server use a custom formatter. It uses the dart_style APIs directly within the analyzer process (which is compiled into the analyzer snapshot).

Once you have a command line tool that can handle formatting requests over stdin/stdout, you can use [the VS Code yeoman templates](https://code.visualstudio.com/api/get-started/your-first-extension) to create an extension (I'd suggest in TypeScript) and I would

- set the activation events to include `onLanguage:dart`
- spawn your process in the extensions `activate` method (note: if you're using a compiled binary, be sure you consider different OSes) and terminate it in the `deactive` method
- Register a formatter using [vscode.languages.registerDocumentFormattingEditProvider](https://code.visualstudio.com/api/references/vscode-api#languages.registerDocumentFormattingEditProvider) that accepts formatting requests from VS Code and sends them to your custom formatter, then creates a `vscode.TextEdit` to update the document

Some other things that might be helpful:

- Dart-Code use a base class used for services that communicate over stdin/stdout [here](https://github.com/Dart-Code/Dart-Code/blob/master/src/shared/services/stdio_service.ts) that might be helpful
- Dart-Code has an implementation of a DocumentFormatteringEditProvider/OnTypeFormattinerProvider [here](https://github.com/Dart-Code/Dart-Code/blob/master/src/extension/providers/dart_formatting_edit_provider.ts) (and registered [here](https://github.com/Dart-Code/Dart-Code/blob/a0bfd8c6606d07579fd6499cb601e35229bfbde0/src/extension/extension.ts#L382)) for the legacy language server integration which shows calling an API to get the formatter text and returning it to VS Code
- You'll probably want to set `dart.enableSdkFormatter: false` when testing because the SDK formatter supports on-type formatting which I'm not sure if VS Code will try to call if your own formatter is not set to handle that yet, which could lead to inconsistent results

**Comment by tekert** [comment]
A Million thanks Dan.

  Before reading your response I was just glazing over [edit_format.dart](https://github.com/dart-lang/sdk/blob/main/pkg/analysis_server/lib/src/handler/legacy/edit_format.dart), and trying to figure out how to make a language server accepting only the minimum for "edit.format" or what vscode required but yep, it's already linked to dart_style, and importing the entire LSP to overwrite one method is not feasible.

  So I was thinking of using an interface for the LSP at https://github.com/devoncarew/analysis_server_lib
I would have to look into the format() method. And try to match the protocol as defined in
[api.html#request_edit.format](https://htmlpreview.github.io/?https://github.com/dart-lang/sdk/blob/master/pkg/analysis_server/doc/api.html#request_edit.format) but haven't tested this approach yet.

  The other option is a standalone bin with a stdin protocol as described, with seems easier to implement, I'm not familiar of how vscode outputs the requests to the server but the code you posted will help me a ton. 
I will look into it and get back after some testing.

**Comment by DanTup** [comment]
The protocol docs you've linked above are for the original Dart analysis server API, which is not a standard VS Code understands (but is instead communicated with using custom code in the Dart extension). That API is no longer used (except in some specific cases, like old Dart SDKs that don't include the LSP server) and I wouldn't recommend building anything on it. For most users everything is done over the LSP protocol (implementation for formatting is at https://github.com/dart-lang/sdk/blob/main/pkg/analysis_server/lib/src/lsp/handlers/handler_formatting.dart).

If your goal is _only_ to provide formatting, then just implementing the format provider in the VS Code and talking to a process with a simple API over stdin/stdout will likely be much simpler than implementing a full LSP server (there's a fair bit of code just handling initialization and capabilities, and communicating using the LSP format, which is like JSON-RPC but with extra headers).

**Comment by tekert** [comment]
Dind't realize, things are changing fast here heh, well, stdin it is, that will accelerate things :)
You mention coding the extension with typescript but out of curiosity, is it feasible to use dart json compiler for an extension? wondering about the performance or documentation.

**Comment by DanTup** [comment]
> You mention coding the extension with typescript but out of curiosity, is it feasible to use dart js compiler for an extension?

In theory, yes. But you'd need to create bindings for the VS Code API (so you can call it from Dart code), and also for any Node API you need (like spawning processes - you can't use `dart:io` with dart2js, so you'd need some bindings to call Node's child_process functions).

I'd previously hoped to write this extension in Dart and compile it that way, but just keeping the VS Code API bindings up-to-date (which there are already typescript definitions for, provided by VS Code) seemed like more effort than it was really worth (and since more and more code is moving out to the SDKs and written in Dart - such as LSP/DP - it seemed less worthwhile).

**Comment by tekert** [comment]
I see, interesting, maybe someday.
And spawning the process using a snapshot is not better for distribution? any reason you mentioned a native bin?

**Comment by DanTup** [comment]
> And spawning the process using a snapshot is not better for distribution? any reason you mentioned a native bin?

Spawning via Dart and using a snapshot would work too, but you might need to locate the `dart` binary. Often people don't have `dart` on `PATH` correctly for VS Code (because they set it in a terminal startup script and then launch VS Code outside of the terminal where it's not inherited). The search logic in this extension is quite tricky, although I suspect you could probably get away with reading the `dart.sdkPath` setting and falling back to just spawning `dart` and assuming it's on `PATH` (and handling failures reasonably).

A native bin avoids that, although you'd have to include one for each platform in the extension (or use [platform-specific vsixs](https://code.visualstudio.com/api/working-with-extensions/publishing-extension#platformspecific-extensions)). Depending on the size of the compiled binary, that might not be a big deal though.
